### PR TITLE
Fix dropdown toggle icon color - RSC-973

### PR DESCRIPTION
**Pull request by rjp99**
https://issues.sonatype.org/browse/RSC-973

### CODE REVIEW

**Comment by entroform** [comment]
Another approach that we saw from nx-tree is to only apply the dark color to nx-icons that is not a threat indicator, `.nx-icon:not(.nx-threat-indicator)`, but we are not sure if there are potentially other icons that uses custom color.

File: gallery/src/components/NxDropdown/NxDropdownPage.tsx [hunk 1]
```diff
@@ -306,6 +319,13 @@ const NxDropdownPage = () =>
       This dropdown contains more complex JSX in its label.
     </GalleryExampleTile>
 
+    <GalleryExampleTile title="Custom label with NxThreatIndicator example"
+                        id="nx-dropdown-with-nx-threat-indicator-example"
+                        liveExample={NxDropdownWithNxThreatIndicatorExample}
+                        codeExamples={nxDropdownWithNxThreatIndicatorExampleCode}>
+      An example of a dropdown with <NxCode>NxThreatIndicator</NxCode>.
```
Comment (rpokorny) [inline gallery/src/components/NxDropdown/NxDropdownPage.tsx:326]: Might want to call out that `nx-dropdown-icon` is _not_ used here and _is_ used in the example above.

File: gallery/src/components/NxDropdown/NxDropdownWithNxThreatIndicatorExample.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import { NxDropdown, useToggle, NxThreatIndicator } from '@sonatype/react-shared-components';
+
+function NxDropdownWithNxThreatIndicatorExample() {
+  const [isOpen, onToggleCollapse] = useToggle(false),
+      onClick = () => { alert('click'); },
+      labelElement = <><NxThreatIndicator threatLevelCategory="moderate" /><span>Moderate</span></>;
+
+  return (
+    <NxDropdown label={labelElement}
+                className="extra-class"
```
Comment (rpokorny) [inline gallery/src/components/NxDropdown/NxDropdownWithNxThreatIndicatorExample.tsx:18]: This seems extraneous to the purpose of the example.

File: gallery/visualtests/nxDropdown.js [hunk 3]
```diff
@@ -90,6 +90,25 @@ describe('NxDropdown', function() {
...
+
+      await button.click();
+    });
+
+    it('looks right', async function() {
+      const [targetElement] = await waitAndGetElements(selector);
+
+      await moveMouseAway();
+
+      await checkScreenshot(targetElement, 251, 252);
```
Comment (rpokorny) [inline gallery/visualtests/nxDropdown.js:107]: It looks like a few px could/should be trimmed off of the bottom of this to minimize the risk of unrelated changes to the documentation page causing diffs.

File: gallery/src/components/NxDropdown/NxDropdownWithNxThreatIndicatorExample.tsx (hunk 2 above)
Comment (rjp99) [reply gallery/src/components/NxDropdown/NxDropdownWithNxThreatIndicatorExample.tsx:18]: Which part? The dropdown example?

Comment (rpokorny) [reply gallery/src/components/NxDropdown/NxDropdownWithNxThreatIndicatorExample.tsx:18]: > Which part? The dropdown example?

The `className`.  It's not doing anything or demonstrating anything, right?

**Comment by rjp99** [comment]
@rpokorny I updated the image size, but wasn't able to run the visual tests or update the visual test with the new image.

**Comment by entroform** [comment]
> > Another approach that we saw from nx-tree is to only apply the dark color to nx-icons that is not a threat indicator, .nx-icon:not(.nx-threat-indicator)
> 
> Where are you saying you saw that code? I'm not finding it. I think when `NxTree` was in-progress it worked like that for a time, but not recently.

Ah sorry, so in nx-tree implementation, it sets the color of nx-icon to grey-05 by default, unless the icon also has the class name `.nx-tree__colored-icon` (`.nx-icon:not(.nx-tree__colored-icon)`).

I was wondering if it's nicer if we do something similar. Instead of creating a new class name for the consuming developers to assign to icons inside nx dropdown, what if by default we set the icon color to grey-05, but make an exception for nx-threat-indicators `.nx-icon:not(.nx-threat-indicator)` (This is assuming nx-threat-indicator is the only icons with custom colors).

**Comment by rpokorny** [comment]
> Ah sorry: So in nx-tree implementation, it sets the color of .nx-icon to grey-05 by default, unless the icon also has the class name .nx-tree__colored-icon (.nx-icon:not(.nx-tree__colored-icon)).
>
> I was wondering if it's nicer if we do something similar. Instead of creating a new class name for the consuming developers to assign to icons inside NxDropdown, what if, by default we set the .nx-icon color to grey-05, but make an exception for nx-threat-indicators .nx-icon:not(.nx-threat-indicator) (This is assuming nx-threat-indicator will be the only icons with custom colors inside NxDropdown).

Ah! So it's doing basically the opposite of what this PR is doing.  Yes, I think we should be consistent about it, and switch this one to be more like that

**Comment by rjp99** [comment]
> > Ah sorry: So in nx-tree implementation, it sets the color of .nx-icon to grey-05 by default, unless the icon also has the class name .nx-tree__colored-icon (.nx-icon:not(.nx-tree__colored-icon)).
> > I was wondering if it's nicer if we do something similar. Instead of creating a new class name for the consuming developers to assign to icons inside NxDropdown, what if, by default we set the .nx-icon color to grey-05, but make an exception for nx-threat-indicators .nx-icon:not(.nx-threat-indicator) (This is assuming nx-threat-indicator will be the only icons with custom colors inside NxDropdown).
> 
> Ah! So it's doing basically the opposite of what this PR is doing. Yes, I think we should be consistent about it, and switch this one to be more like that

We made the change here:
f0f6966a017d1b7b89c1932d3391193950aa2935

**Comment by rpokorny** [comment]
> We made the change here:
https://github.com/sonatype/sonatype-react-shared-components/commit/f0f6966a017d1b7b89c1932d3391193950aa2935

Oh, it is important IMO though that it be a new class just for this purpose, rather than writing the styles specifically for `.nx-threat-indicator`.  We don't want to make the assumption that that will be the only colored icon that might ever get used here.

**Comment by entroform** [comment]
> > We made the change here:
> > [f0f6966](https://github.com/sonatype/sonatype-react-shared-components/commit/f0f6966a017d1b7b89c1932d3391193950aa2935)
> 
> Oh, it is important IMO though that it be a new class just for this purpose, rather than writing the styles specifically for `.nx-threat-indicator`. We don't want to make the assumption that that will be the only colored icon that might ever get used here.

Hmm, what are your thoughts on creating `nx-icon--colored` modifier and setting that inside `NxThreatIndicator` icon and other colored icons in the future? We can document that modifier nx-icon documentation and note it in NxTree and NxDropdown just incase devs want to create custom colored icons, but it is set by default for all RSC colored icons. It could be a better DX for consuming developers, so they don't have to worry about applying a classname to RSC components when they are being composed.

**Comment by rpokorny** [comment]
> Hmm, what are your thoughts on creating nx-icon--colored modifier and setting that inside NxThreatIndicator icon and other colored icons in the future? We can document that modifier nx-icon documentation and note it in NxTree and NxDropdown just incase devs want to create custom colored icons, but it is set by default for all RSC colored icons. It could be a better DX for consuming developers, so they don't have to worry about applying a classname to RSC components when they are being composed.

That sounds like a decent idea but I'd say doing that properly is significantly out of scope for this ticket - a new documentation page, modifying `NxTree`, figuring out what to do about the existing approach in terms of breaking changes, etc.  Let's do it the existing way for now and then as either another ticket or on Improvement Day someone can explore the approach that you're suggesting.

**Review by rpokorny — APPROVED** [review]

File: gallery/src/components/NxDropdown/NxDropdownPage.tsx (hunk 1 above)
Comment (entroform) [reply gallery/src/components/NxDropdown/NxDropdownPage.tsx:327]: > Might want to call out that `nx-dropdown-icon` is _not_ used here and _is_ used in the example above.

We added it here: f7052b6057df0e5d4a37409ddf30b9e601d5be42

**Comment by entroform** [comment]
@rpokorny Hi Ross, we went back to our original implementation, (using `nx-dropdown-icon` ) because we noticed that it was overriding disabled state icon buttons, the reset button on dropdown with filter, and external link icon color. After this discovery, we thought the original solution might be the simpler way to go. Let me know what you think!

**Comment by rpokorny** [comment]
> @rpokorny Hi Ross, we went back to our original implementation, (using nx-dropdown-icon ) because we noticed that it was overriding disabled state icon buttons, the reset button on dropdown with filter, and external link icon color. After this discovery, we thought the original solution might be the simpler way to go. Let me know what you think!

That's fine I guess. Should it be `nx-dropdown-icon` or something like `nx-dropdown__icon` though?

File: gallery/src/components/NxDropdown/NxDropdownPage.tsx [hunk 4]
```diff
@@ -258,6 +260,16 @@ const NxDropdownPage = () =>
                 any <NxCode>.nx-dropdown-right-button</NxCode>s.
               </NxTable.Cell>
             </NxTable.Row>
+            <NxTable.Row>
+              <NxTable.Cell><NxCode>nx-dropdown-icon</NxCode></NxTable.Cell>
+              <NxTable.Cell>
+                Set custom icons to the correct color.
+              </NxTable.Cell>
+              <NxTable.Cell>
+                This class should be applied to custom icons that are not colored{' '}
```
Comment (rpokorny) [inline gallery/src/components/NxDropdown/NxDropdownPage.tsx:269]: "are not colored" might not be the clearest phrasing. Maybe "do not have a built-in color"

**Review by rpokorny — APPROVED** [review]

File: gallery/src/components/NxDropdown/NxDropdownPage.tsx (hunk 4 above)
Comment (entroform) [reply gallery/src/components/NxDropdown/NxDropdownPage.tsx]: > "are not colored" might not be the clearest phrasing. Maybe "do not have a built-in color"

Updated: 34e67258d5977d2fd3d77619b29f01a416f2bfaf
